File: 239.滑动窗口最大值.py
#!/usr/bin/env python
# coding=utf-8
'''
Description: 
Author: yangyuxiang
Date: 2021-05-11 08:17:35
LastEditors: yangyuxiang
LastEditTime: 2021-06-28 09:11:17
FilePath: /leetcode/239.滑动窗口最大值.py
'''
#
# @lc app=leetcode.cn id=239 lang=python
#
# [239] 滑动窗口最大值
#

# @lc code=start
import collections

class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        queue = collections.deque()
        for i in range(k):
            while queue and nums[i] >= nums[queue[-1]]:
                queue.pop()
            queue.append(i)

        ans = [nums[queue[0]]]
        for i in range(k, len(nums)):
            if i - k >= queue[0]:
                queue.popleft()
            while queue and nums[i] >= nums[queue[-1]]:
                queue.pop()
            queue.append(i)

            ans.append(nums[queue[0]])

        return ans


if __name__ == '__main__':
    nums = [1,3,-1,-3,5,3,6,7]
    k = 3
    print(Solution().maxSlidingWindow(nums, k))
             

            


        
# @lc code=end

File: 239.滑动窗口最大值.java
import java.util.Deque;
import java.util.LinkedList;


/*
 * @lc app=leetcode.cn id=239 lang=java
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
class MaxSlidingWindow {
    public int[] maxSlidingWindow(int[] nums, int k) {
        // 单调递减队列
        if (nums.length <= 1) {
            return nums;
        }
        Deque<Integer> deque = new LinkedList<>();  // 存放索引
        int[] res = new int[nums.length - k + 1];
        int num = 0;
        for (int i = 0; i < k; i++) {
            while (!deque.isEmpty() && nums[deque.getLast()] < nums[i]) {
                deque.removeLast();
            }
            deque.addLast(i);
        }
        res[num++] = nums[deque.peek()];
        for (int i=k; i<nums.length; i++){
            if (i - k >= deque.peek()) {  // 判断单调队列的首元素是否需要出列
                deque.removeFirst();
            }
            while (!deque.isEmpty() && nums[deque.getLast()] < nums[i]) {
                deque.removeLast();
            }
            deque.addLast(i);
            res[num++] = nums[deque.peek()];
        }
        return res;
    }

    public static void main(String[] args) {
        int[] nums = {1,3,-1,-3,5,3,6,7};
        int k = 3;
        System.out.println(new MaxSlidingWindow().maxSlidingWindow(nums, k));
    }
}
// @lc code=end

